### PR TITLE
feat(pace-52): add evaluate command to apply a policy to sample csv data

### PR DESCRIPTION
**Pull request by ivan-p92**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,8 +3,8 @@ module pace/pace
 go 1.20
 
 require (
-	buf.build/gen/go/getstrm/pace/grpc/go v1.3.0-20231121141154-e945a7aece2d.2
-	buf.build/gen/go/getstrm/pace/protocolbuffers/go v1.28.1-20231121141154-e945a7aece2d.4
+	buf.build/gen/go/getstrm/pace/grpc/go v1.3.0-20231122150633-2d7e63f278de.2
```
Comment (ivan-p92) [inline go.mod:6]: Reminder: use eventual latest version

Comment (trietsch) [reply go.mod:6]: Eventual consistent latest maybe version.

File: pkg/entity/datapolicy/cmd.go [hunk 2]
```diff
@@ -55,6 +55,34 @@ func ApplyCmd() *cobra.Command {
...
+			evaluate(cmd, &args[0])
+		},
+		Args:              cobra.ExactArgs(1), // the policy id
+		ValidArgsFunction: IdsCompletion,
+	}
+
+	flags := cmd.Flags()
+	processingplatform.AddProcessingPlatformFlag(cmd, flags)
+	cmd.MarkFlagRequired(common.ProcessingPlatformFlag)
+	addSampleDataFlag(cmd, flags)
```
Comment (trietsch) [inline pkg/entity/datapolicy/cmd.go:78]: This one can move in place IMO. We're not reusing it anywhere (yet).

File: pkg/entity/datapolicy/datapolicy.go [hunk 3]
```diff
@@ -143,6 +160,16 @@ func readPolicy(filename string) *DataPolicy {
 	return dataPolicy
 }
 
+func addSampleDataFlag(cmd *cobra.Command, flags *pflag.FlagSet) {
+	flags.String(common.SampleDataFlag, "", common.SampleDataUsage)
+	CliExit(
+		cmd.RegisterFlagCompletionFunc(common.SampleDataFlag,
+			func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+				return []string{"csv,txt"}, cobra.ShellCompDirectiveFilterFileExt
```
Comment (trietsch) [inline pkg/entity/datapolicy/datapolicy.go:168]: Why complete on `txt`? Do you think that many people would suffix their `csv` content with `txt` in the filename?

File: pkg/entity/datapolicy/texts.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ Applies an existing data policy to the target platform.
 `)
 
 var applyExample = LongDocs(`
-!pace apply data-policy public.demo
+!pace apply data-policy public.demo --processing-platform bigquery-dev
```
Comment (trietsch) [inline pkg/entity/datapolicy/texts.go:42]: Good catch.

File: pkg/entity/datapolicy/texts.go [hunk 5]
```diff
@@ -59,6 +59,46 @@ data_policy:
 ...
 `)
 
+var evaluateLongDocs = LongDocs(`
+Evaluates an existing data policy by applying it to sample data provided in a csv file.
+You can use this to verify the correctness of your field transforms and filters.
```
Comment (trietsch) [inline pkg/entity/datapolicy/texts.go:64]: Verify is not true, as we cannot use all transforms.

```suggestion
You can use this to test the correctness of your field transforms and filters.
```

File: pkg/entity/datapolicy/texts.go [hunk 6]
```diff
@@ -59,6 +59,46 @@ data_policy:
...
+The csv file should contain a header row with the column names, matching the fields in the data policy.
+A comma should be used as the delimiter.
```
Comment (trietsch) [inline pkg/entity/datapolicy/texts.go:66]: This leads me to think that the `CSV` sample data in the proto should be more configurable. Maybe not now, but later. Therefore, the csv should not be a `string` in the proto field, but a message.

File: pkg/entity/datapolicy/texts.go [hunk 7]
```diff
@@ -59,6 +59,46 @@ data_policy:
...
+Currently, only standard SQL data types are supported. For platform-specific transforms, test on the platform itself.
+`)
+
+var evaluateExample = LongDocs(`
+!pace evaluate data-policy public.demo --processing-platform example-platform --sample-data sample.csv
```
Comment (trietsch) [inline pkg/entity/datapolicy/texts.go:71]: This is awesome!

File: pkg/util/utils.go [hunk 8]
```diff
@@ -70,9 +70,14 @@ func CliExit(err error) {
 		st, ok := status.FromError(err)
 
 		if ok {
-			details := st.Details()[0]
-			yamlBytes := ProtoMessageToYaml(details.(proto.Message))
-			additionalDetails := string(yamlBytes.Bytes())
+			var additionalDetails string
+			if len(st.Details()) > 0 {
```
Comment (trietsch) [inline pkg/util/utils.go:74]: Yup, forgot that check, thanks!

**Review by trietsch — CHANGES_REQUESTED** [review]

File: pkg/entity/datapolicy/cmd.go (hunk 2 above)
Comment (ivan-p92) [reply pkg/entity/datapolicy/cmd.go:78]: It's a bit verbose, unless you meant in this file as a function?

**Pull request merged**